feat: revamp charm logging and status-handling

pedro-avalos opened the pull request:
This PR improves the logging and status messages of the charm.

Additionally, this PR replaces the use of `subprocess` with the `operator_libs_linux.systemd` charm library.

pedro-avalos committed:
chore: improve logging for _on_install

pedro-avalos committed:
chore: improve logging in _on_start

pedro-avalos committed:
chore: improve logging for _on_config_changed

pedro-avalos committed:
chore: add gitignore

rene-oromtz commented on src/charm.py:
```diff
@@ -38,80 +40,83 @@ def _on_install(self, _):
+        try:
+            decoded_config = base64.b64decode(config_content).decode("utf-8")
+        except binascii.Error as e:
+            logging.exception("Failed to decode config-file content: %s", e)
```
```suggestion
            logger.exception("Failed to decode config-file content: %s", e)
```

rene-oromtz commented on src/charm.py:
```diff
@@ -121,7 +126,8 @@ def _on_start(self, _):
             self.unit.status = ops.ActiveStatus()
-        except subprocess.CalledProcessError:
+        except subprocess.CalledProcessError as e:
+            logger.exception("Failed to start Conserver service: %s", e)
```
I think there is a minor issue during install that we could address. 
For some reason, seems that even when the `_on_config_changed` runs before the `_on_start`, this second hook is already "queued" so it ignores the blocked status. Since the config file is required by conserver, it will ended up "Blocked" but for the incorrect reason `Service failed to start` instead of `Missing config-file in config`. 

Maybe we can remove this status check out from this hook, similarly as the approach done by the[ snmp-exporter charm](https://github.com/canonical/snmp-exporter-operator/blob/5ea7d179ab35e54cef0b6c982cb2c030631a3873/src/charm.py#L111) where configuration files are also required. 

Logs:
```
unit-conserver2-0: 09:48:56 ERROR unit.conserver2/0.juju-log config-file not found in charm configuration
unit-conserver2-0: 09:48:56 INFO juju.worker.uniter.operation ran "config-changed" hook (via hook dispatching script: dispatch)
unit-conserver2-0: 09:48:56 INFO juju.worker.uniter found queued "start" hook
unit-conserver2-0: 09:48:57 INFO unit.conserver2/0.juju-log Running legacy hooks/start.
unit-conserver2-0: 09:48:57 ERROR unit.conserver2/0.juju-log Failed to start Conserver service: Command '['systemctl', 'is-active', '--quiet', 'conserver-server']' returned non-zero exit status 3.
Traceback (most recent call last):
  File "/var/lib/juju/agents/unit-conserver2-0/charm/./src/charm.py", line 125, in _on_start
    subprocess.check_call(
  File "/usr/lib/python3.10/subprocess.py", line 369, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['systemctl', 'is-active', '--quiet', 'conserver-server']' returned non-zero exit status 3.

```

rene-oromtz replied:
Or another simpler approach is not let the status check run if the unit status was previously blocked:
```

def _on_start(self, _):
        """Handle start event."""
        if isinstance(self.unit.status, ops.BlockedStatus):
            return
        try:
            # Ensure service is running
            ...

```

pedro-avalos replied:
I have revamped the way the status-handling logic is done, extracted it all into a `set_status` function so that we can ensure we set the right status no matter where we are in the charm.

rene-oromtz reviewed (changes requested):
Thanks for taking care of this! 
I added some comments due to a minor issue that will only happen during initial deployment if no config was set, please let me know what you think

pedro-avalos committed:
fix: use logger, not logging directly

pedro-avalos committed:
feat: add operator_libs_linux.systemd charm-lib

pedro-avalos committed:
feat: extract _write_file

this simplifies the logic to write either the conserver.cf or conserver.passwd files

pedro-avalos committed:
feat: use systemd charm-lib

this commit also adds a set_status helper to ensure the right status is used regardless of which hook is executed/queued last

pedro-avalos committed:
feat: add stop event

pedro-avalos committed:
chore: move helper function to bottom

pedro-avalos committed:
chore: fix linting

rene-oromtz reviewed (approved):
Thanks for addressing the changes, this looks really good 👌
